File: apps/web-asset/src/api/goals.ts
import { http } from "./http";

export type GoalScopeType = "USER" | "HOUSEHOLD";
export type GoalBasis = "GROSS" | "NET";
export type GoalDisplayCurrency = "KRW" | "USD";

export type GoalTargetOut = {
  configured: boolean;
  scope_type: GoalScopeType;
  scope_id: number;
  display_currency: GoalDisplayCurrency;
  amount_currency: GoalDisplayCurrency;
  target_amount: string | number | null;
  annual_return_rate_pct: string | number | null;
  monthly_invest_amount: string | number | null;
  created_at: string | null;
  updated_at: string | null;
};

export type GoalTargetUpdateIn = {
  scope_type: GoalScopeType;
  scope_id?: number;
  display_currency: GoalDisplayCurrency;
  target_amount: string | number;
  annual_return_rate_pct: string | number;
  monthly_invest_amount: string | number;
};

export type AnalyticsGoalProgressOut = {
  configured: boolean;
  scope_type: GoalScopeType;
  scope_id: number;
  basis: GoalBasis;
  display_currency: GoalDisplayCurrency;
  current_amount: string | number;
  target_amount: string | number | null;
  progress_ratio_pct: string | number | null;
  remaining_amount: string | number | null;
  over_target_amount: string | number | null;
  reached: boolean;
  projected_reach_date: string | null;
  projected_months_to_goal: number | null;
  projection_3y: string | number | null;
  projection_5y: string | number | null;
  projection_10y: string | number | null;
  recent_actual_annualized_return_pct: string | number | null;
  recent_actual_window_days: number | null;
  comparison_tone: "AHEAD" | "BEHIND" | "MATCHED" | "UNAVAILABLE";
  as_of: string;
};

export type GoalTargetQuery = {
  scope_type: GoalScopeType;
  scope_id?: number;
  display_currency: GoalDisplayCurrency;
};

export type GoalProgressQuery = GoalTargetQuery & {
  basis?: GoalBasis;
};

export async function getMyGoalTarget(params: GoalTargetQuery): Promise<GoalTargetOut> {
  const { data } = await http.get<GoalTargetOut>("/users/me/goal-target", { params });
  return data;
}

export async function updateMyGoalTarget(payload: GoalTargetUpdateIn): Promise<GoalTargetOut> {
  const { data } = await http.put<GoalTargetOut>("/users/me/goal-target", payload);
  return data;
}

export async function getGoalProgress(params: GoalProgressQuery): Promise<AnalyticsGoalProgressOut> {
  const { data } = await http.get<AnalyticsGoalProgressOut>("/analytics/goal-progress", { params });
  return data;
}

File: apps/api/app/schemas/goals.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


ScopeType = Literal["USER", "HOUSEHOLD"]
DisplayCurrency = Literal["KRW", "USD"]
GoalBasis = Literal["GROSS", "NET"]
ComparisonTone = Literal["AHEAD", "BEHIND", "MATCHED", "UNAVAILABLE"]


class GoalTargetOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    configured: bool
    scope_type: ScopeType
    scope_id: int
    display_currency: DisplayCurrency
    amount_currency: DisplayCurrency
    target_amount: Decimal | None = None
    annual_return_rate_pct: Decimal | None = None
    monthly_invest_amount: Decimal | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class GoalTargetUpdateIn(BaseModel):
    scope_type: ScopeType
    scope_id: int | None = None
    display_currency: DisplayCurrency
    target_amount: Decimal = Field(gt=0)
    annual_return_rate_pct: Decimal = Field(ge=0, le=100)
    monthly_invest_amount: Decimal = Field(ge=0)


class AnalyticsGoalProgressOut(BaseModel):
    configured: bool
    scope_type: ScopeType
    scope_id: int
    basis: GoalBasis
    display_currency: DisplayCurrency
    current_amount: Decimal
    target_amount: Decimal | None = None
    progress_ratio_pct: Decimal | None = None
    remaining_amount: Decimal | None = None
    over_target_amount: Decimal | None = None
    reached: bool
    projected_reach_date: date | None = None
    projected_months_to_goal: int | None = None
    projection_3y: Decimal | None = None
    projection_5y: Decimal | None = None
    projection_10y: Decimal | None = None
    recent_actual_annualized_return_pct: Decimal | None = None
    recent_actual_window_days: int | None = None
    comparison_tone: ComparisonTone
    as_of: datetime
